core/src/com/unciv/logic/automation/unit/EspionageAutomation.kt

```diff
@@ -1,33 +1,83 @@
-                spy.moveTo(getCivsToStealFromSorted.first().cities.filter { it.getCenterTile().isVisible(civInfo) }.maxByOrNull { it.population.population })
+    fun automateSpies() {
+        val spies = civInfo.espionageManager.spyList
+        val spiesDoingCounterIntelligence = spies.count { it.action == SpyAction.CounterIntelligence }
+        val spiesToMove = spies.filter { it.isAlive() && 
+            (!it.isDoingWork() || (it.action == SpyAction.CounterIntelligence && spiesDoingCounterIntelligence > 2))}
+        for (spy in spiesToMove) {
+            val randomSeed = spies.size + spies.indexOf(spy) + civInfo.gameInfo.turns
+            val randomAction = Random(randomSeed).nextInt(10)
+
+            // Try each operation based on the random value and the success rate
+            // If an operation was not successfull try the next one
+            if (randomAction <= 7 && automateSpyStealTech(spy)) {
                 continue
+            } else if (randomAction <= 9 && automateSpyRigElection(spy)) {
+                continue
+            } else if (automateSpyCounterInteligence(spy)) {
+                continue
+            } else if (spy.isDoingWork()) {
+                continue // We might have been doing counter intelligence and wanted to look for something better
+            } else {
+                // Retry all of the operations one more time
+                if (automateSpyStealTech(spy)) continue
+                if (automateSpyRigElection(spy)) continue
+                if(automateSpyCounterInteligence(spy)) continue
             }
-            if (spy.action == SpyAction.None) {
-                spy.moveTo(civInfo.getKnownCivs().filter { otherCiv -> otherCiv.isMajorCiv() && otherCiv.cities.any { it.getCenterTile().isVisible(civInfo) }}
-                    .toList().randomOrNull()?.cities?.filter { it.getCenterTile().isVisible(civInfo) }?.randomOrNull())
-            }
+            // There is nothing for our spy to do, put it in a random city
+            spy.moveTo(civInfo.gameInfo.getCities().filter { it.getCenterTile().isVisible(civInfo) && spy.canMoveTo(it) }.toList().randomOrNull())
         }
     }
+
+    /**
+     * Moves the spy to a city that we can steal a tech from
+     */
+    fun automateSpyStealTech(spy: Spy): Boolean {
+        if (civsToStealFrom.isNotEmpty()) {
+            // We want to move the spy to the city with the highest science generation
+            // Players can't usually figure this out so lets do highest population instead
+            spy.moveTo(getCivsToStealFromSorted.first().cities.filter { spy.canMoveTo(it) }.maxByOrNull { it.population.population })
+            return true
+        }
+        return false
+    }
+
+    /**
+     * Moves the spy to a random city-state
+     */
+    private fun automateSpyRigElection(spy: Spy): Boolean {
+        val potentialCities = cityStatesToRig.flatMap { it.cities }.filter { !it.isBeingRazed && spy.canMoveTo(it) }
```

Comment:
Just because we know them, doesn't mean we know the location of all their cities - can we move to a city we don't know of yet?

Reply:
Spy.canMoveTo(city) checks if we can see the city center tile. It cleans up a lot of the code here.